Fix tfs push bug

smile21prc opened the pull request:
The fix is ready, which basically:

1. Handles the case when SCM configuration of a job is empty, which was the
root cause of the previous random TFS push trigger failure.

2. Add an additional case when a job should be executed – the commit is NOT
null:
a. runPolling() actually checks SCM changes, which in our cases are
none.
b. As long as a commit in linked repro triggers this push (commit is NOT
null), the job should be executed.

3. Refactors original trigger logic into function “triggerJob”, to avoid
dupe code as these code will be called twice in AbstractHookEvent.java.

4. I removed existing code in AbstractHookEvent.java, which goes through all SCMs and Repos and might still be applied to some old free style projects. This is because its logic is dupe with the one in runPolling(). 

Now if you merge any change to dotnet-linker/enable-ci, all four jobs
starting with “testing” (on link below) will be triggered, and ONLY
those three will be triggered.

https://dotnet-vsts.westus2.cloudapp.azure.com/job/DevDiv_dotnet-linker/job/enable-ci/

smile21prc committed:
Fix tfs push bug

The fix is ready, which basically:

Handles the case when SCM configuration of a job is empty, which was the
root cause of the previous random TFS push trigger failure.
Add an additional case when a job should be executed – the commit is NOT
null:
a. runPolling() actually checks SCM changes, which in our cases are
none.
b. As long as a commit in linked repro triggers this push (commit is NOT
null), the job should be executed.
Refactors original trigger logic into function “triggerJob”, to avoid
dupe code as these code will be called twice in AbstractHookEvent.java.
I choose not to remove existing code in AbstractHookEvent.java, which
goes through all SCMs and Repos and might still be applied to some old
free style projects. Instead I add the handling of the missing cases
(when SCM configuration of a job is empty) on top of it.

Now if you merge any change to dotnet-linker/enable-ci, all four jobs
starting with “testing” (on link below) will be triggered, and ONLY
those three will be triggered.

https://dotnet-vsts.westus2.cloudapp.azure.com/job/DevDiv_dotnet-linker/job/enable-ci/

smile21prc committed:
Cleanup

Cleanup

smile21prc committed:
Still need this iteration

Still need this iteration to filter out unwanted runs.

smile21prc committed:
Clean up spaces

Clean up spaces

smile21prc committed:
Revert "Clean up spaces"

This reverts commit 38d1c0ffc92a12e3b7c72ff349c616e1a608b428.

smile21prc committed:
Revert "Still need this iteration"

This reverts commit c3454981fa2af9502cdad6095a3528bcd79ec75c.

smile21prc committed:
Remove unnecessary continue

Remove unnecessary continue

jpricket commented on tfs/src/main/java/hudson/plugins/tfs/TeamPushTrigger.java:
```diff
@@ -119,13 +119,13 @@ public void run() {
             String changesDetected = "";
             if (!bypassPolling) {
-                if (runPolling()) {
+                if (runPolling() || gitCodePushedEventArgs.commit != null) {
```
I would prefer...
|| StringUtils.isNotBlank(gitCodePushedEventArgs.commit)

smile21prc replied:
Good suggestion Jason, I'll fix and verify it.

smile21prc committed:
Address review comments.

Address review comments.

jpricket commented:
@smile21prc - I few things about process:
1) Please squash your commits (this will make history easier to follow)
2) Add appropriate Unit Tests for any new features (I don't think you need any for this change)
3) To create a PR against the tfs-plugin repo, push your commit to your master branch in this repo and create a PR against the master branch in the tfs-plugin repo. This will create a PR in the other repo that can then be merged.
4) Make sure that the Jenkins builds pass. If not, please correct any issues and update the master branch here with the fixes. This will automatically update the PR.

Thanks for this contribution!

jpricket commented:
@smile21prc - when the new PR is ready in the tfs-plugin branch, I will review it and merge it when ready.

smile21prc commented:
@jpricketMSFT , thanks for the review and instructions about the process. I'll follow your instructions and let you know when the new PR is ready.

smile21prc committed:
Add import of StringUtils.

Add import of StringUtils.

smile21prc commented:
@mmitche , I've verified our jobs can constantly be triggered now by any push in enable-ci branch after addressing Jason's comment. Any additional comment before I merge into my private master branch?

mmitche commented:
@smile21prc I wouldn't merge this into your private master.  This should just be a PR to jenkinsci/tfs-plugin.

I'll take a look at this later.

smile21prc commented:
Thanks @mmitche , I'll hold it till you take a look.

mmitche commented on tfs/src/main/java/hudson/plugins/tfs/model/AbstractHookEvent.java:
```diff
@@ -81,6 +81,68 @@ static JSONObject fromResponseContributors(final List<GitStatus.ResponseContribu
+        
+        if (!(project instanceof AbstractProject && ((AbstractProject) project).isDisabled())) {
+            if (project instanceof Job) {
+                // TODO: Add default parameters defined in the job
```
Is it really not queueing it with parameters?

smile21prc replied:
Verified the jobs being kicked off do use customized params, so this comment isn't needed any more, removed it.

mmitche commented on tfs/src/main/java/hudson/plugins/tfs/model/AbstractHookEvent.java:
```diff
@@ -81,6 +81,68 @@ static JSONObject fromResponseContributors(final List<GitStatus.ResponseContribu
+                final int quietPeriod = scmTriggerItem.getQuietPeriod();                
+                
+                boolean triggered = false;
+                if (!triggered) {
```
This is always false here

smile21prc replied:
Yep, this is the code from original plugin. I've removed this unnecessary if statement.

mmitche commented on tfs/src/main/java/hudson/plugins/tfs/model/AbstractHookEvent.java:
```diff
@@ -81,6 +81,68 @@ static JSONObject fromResponseContributors(final List<GitStatus.ResponseContribu
+                        final SCMTrigger scmTrigger = TeamEventsEndpoint.findTrigger(job, SCMTrigger.class);
+                        if (scmTrigger != null && scmTrigger.isIgnorePostCommitHooks()) {
+                            // job has explicitly opted out of hooks
+                            triggered = false;
```
There is a lot of use of triggered in here in sort of confusing ways.  I think the if (triggered) block below should be inlined into this block (invert the above if condition), returning the new gitResponse directly rather than waiting for the end of the series of if statements.

smile21prc replied:
Refined the logic, and removed "triggered" var entirely as it becomes unnecessary. Also changed to return directly instead of waiting.

mmitche commented on tfs/src/main/java/hudson/plugins/tfs/model/AbstractHookEvent.java:
```diff
@@ -81,6 +81,68 @@ static JSONObject fromResponseContributors(final List<GitStatus.ResponseContribu
+                        }
+                    }
+                }
+                if (!triggered) {
```
If you inline the if (triggered) block, then this if (!triggered) becomes unnecessary since it's always false if you get here.

smile21prc replied:
Yep, removed "triggered" var entirely as it becomes unnecessary.

mmitche commented on tfs/src/main/java/hudson/plugins/tfs/model/AbstractHookEvent.java:
```diff
@@ -81,6 +81,68 @@ static JSONObject fromResponseContributors(final List<GitStatus.ResponseContribu
+                        final Action[] actionArray = ActionHelper.create(actions, causeAction);
+                        scmTriggerItem.scheduleBuild2(quietPeriod, actionArray);
+                        gitResponse = new TeamEventsEndpoint.ScheduledResponseContributor(project);
+                        triggered = true;
```
Again you could just return gitResponse directly from this block

smile21prc replied:
Good suggestion, did that.

mmitche commented on tfs/src/main/java/hudson/plugins/tfs/model/AbstractHookEvent.java:
```diff
@@ -81,6 +81,68 @@ static JSONObject fromResponseContributors(final List<GitStatus.ResponseContribu
+                        triggered = true;
+                    }
+                }
+                if (!triggered) {
```
Which would make this always false again.

smile21prc replied:
Agree, removed this unnecessary statement.

mmitche commented on tfs/src/main/java/hudson/plugins/tfs/model/AbstractHookEvent.java:
```diff
@@ -81,6 +81,68 @@ static JSONObject fromResponseContributors(final List<GitStatus.ResponseContribu
+                        else {
+                            gitResponse = new TeamEventsEndpoint.PollingScheduledResponseContributor(project);
+                        }
+                        triggered = true;
```
And again return directly from this

smile21prc replied:
Done.

mmitche commented on tfs/src/main/java/hudson/plugins/tfs/TeamPushTrigger.java:
```diff
@@ -119,13 +120,13 @@ public void run() {
             String changesDetected = "";
             if (!bypassPolling) {
-                if (runPolling()) {
+                if (runPolling() || StringUtils.isNotBlank(gitCodePushedEventArgs.commit)) {
```
Can you add a comment here to indicate why the gitCodePushedEventArgs is needed?

smile21prc replied:
Sure, added comments below:
                // pipeline jobs might have runPolling() returned as false, while still should be scheduled.
                // we should shedule them as long as they are associated with a valid commit.

mmitche replied:
Any idea why runPolling returns false in those cases?

mmitche replied:
(and if so, can you add it as a comment here too)

smile21prc commented on tfs/src/main/java/hudson/plugins/tfs/model/AbstractHookEvent.java:
```diff
@@ -81,6 +81,68 @@ static JSONObject fromResponseContributors(final List<GitStatus.ResponseContribu
+        
+        if (!(project instanceof AbstractProject && ((AbstractProject) project).isDisabled())) {
+            if (project instanceof Job) {
+                // TODO: Add default parameters defined in the job
```
Seems that's what's be doing now. If there is a way to enhance it, please let me know and I'll apply that.

smile21prc committed:
Addressed Matt's comments.

Addressed Matt's comments.

smile21prc committed:
Addressed more of Matt's comments.

Addressed more of Matt's comments.

smile21prc commented:
Addressed all comments, and re-created this PR as below against the branch to be merged (master in jenkinsci/tfs-plugin)
https://github.com/jenkinsci/tfs-plugin/pull/164
